### PR TITLE
Allow `SmartAllocator` to be used with arbitary allocators.

### PR DESCRIPTION
**Pull request by Lymia**

### CODE REVIEW

File: src/smart.rs [hunk 1]
```diff
@@ -1,29 +1,32 @@
...
 /// for all given memory types.
 ///
 /// Allocates memory blocks from the least used memory type from those which satisfy requirements.
 #[derive(Debug)]
-pub struct SmartAllocator<B: Backend> {
-    allocators: Vec<(MemoryType, CombinedAllocator<B>)>,
+pub struct SmartAllocator<B: Backend, A: MemoryAllocator<B> = CombinedAllocator<B>> {
+    allocators: Vec<(MemoryType, A)>,
     heaps: Vec<Heap>,
+    _phantom: PhantomData<fn(B) -> B>,
```
Comment (zakarumych) [inline src/smart.rs]: `fn(B)` or `fn() -> B` is enough though 😄

File: src/smart.rs [hunk 2]
```diff
@@ -38,30 +41,46 @@ where
...
+    A: MemoryAllocator<B>,
+{
+    /// Create a new smart allocator from `MemoryProperties` given by a device, backed by a
+    /// custom allocator.
+    ///
+    /// ### Parameters:
+    ///
+    /// - `memory_properties`: memory properties describing the memory available on a device
+    /// - `new_allocator`: the function used to create an allocator for each memory type
+    pub fn for_allocator<F: FnMut(MemoryTypeId) -> A>(
```
Comment (zakarumych) [inline src/smart.rs]: `with_allocator`. Or just plane `new` since there is no `new` method.

File: src/smart.rs [hunk 3]
```diff
@@ -181,13 +201,16 @@ impl Heap {
 
 /// `Block` type returned by `SmartAllocator`.
 #[derive(Debug)]
-pub struct SmartBlock<M>(CombinedBlock<M>, usize);
+pub struct SmartBlock<M: Debug + Any, B: Block<Memory = M> = CombinedBlock<M>>(
```
Comment (zakarumych) [inline src/smart.rs]: Why `Any`?

Comment (zakarumych) [reply src/smart.rs]: Why `M` at all. You are able to calculate it from `B` alone.

**Review by zakarumych — CHANGES_REQUESTED** [review]

File: src/smart.rs (hunk 2 above)
Comment (Lymia) [reply src/smart.rs]: There is a new that applies only to `SmartAllocator<B>` (compatible with the older API.

File: src/smart.rs (hunk 3 above)
Comment (Lymia) [reply src/smart.rs]: The idea is to keep the old usage of `SmartAllocator<B>`, so it can still be used as such. Is this unnecessary?

Comment (zakarumych) [reply src/smart.rs]: I think it's better to use type-aliases instead.

File: src/smart.rs [hunk 4]
```diff
@@ -1,89 +1,76 @@
...
+use combined::{CombinedAllocator, CombinedBlock};
 
 /// Allocator that can choose memory type based on requirements, and keeps track of allocators
 /// for all given memory types.
 ///
 /// Allocates memory blocks from the least used memory type from those which satisfy requirements.
 #[derive(Debug)]
-pub struct SmartAllocator<B: Backend> {
-    allocators: Vec<(MemoryType, CombinedAllocator<B>)>,
+pub struct GenericSmartAllocator<B: Backend, A: MemoryAllocator<B>> {
```
Comment (zakarumych) [inline src/smart.rs:16]: Now `B` is not necessary here.

**Review by zakarumych — CHANGES_REQUESTED** [review]

**Comment by zakarumych** [comment]
Otherwise, great work!

File: src/smart.rs (hunk 4 above)
Comment (Lymia) [reply src/smart.rs:16]: `B` is necessary, as `MemoryAllocator` has a type parameter, and there is no other way to have a type parameter bounded on a trait with a type parameter.

The only other option is to define `SmartAllocator` as having no trait bounds on its type parameters at all, and have it solely on the impl bounds.

Comment (zakarumych) [reply src/smart.rs:16]: The only option seems good.

Comment (Lymia) [reply src/smart.rs:16]:
```
error[E0207]: the type parameter `B` is not constrained by the impl trait, self type, or predicates
  --> src/smart.rs:21:7
   |
21 | impl <B, A> GenericSmartAllocator<A>
   |       ^ unconstrained type parameter
```

.. it's actually not an option, unless I'm missing a fix to this.

Comment (zakarumych) [reply src/smart.rs:16]: Can you just remove `B` from there?

Comment (zakarumych) [reply src/smart.rs:16]: If you need `B` in functions - add it to the functions

Comment (zakarumych) [reply src/smart.rs:16]: For trait implementation `impl<B, A> MemoryAllocator<B> for GenericSmartAllocator<A>` should work

**Pull request closed without merging**